swim/handlers_test.go

```diff
@@ -0,0 +1,222 @@
+	s.NoError(err, "calling handler should not result in error")
+
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestToggleGossipHandler() {
+	s.Require().True(s.testNode.node.gossip.Stopped())
+
+	var err error
+
+	// Toggle gossip on
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.False(s.testNode.node.gossip.Stopped())
+
+	// Toggle gossip off
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestAdminLeaveJoinHandlers() {
+	// Join test node to cluster
+	s.cluster.Add(s.testNode.node)
+	s.Require().Equal(5, s.testNode.node.CountReachableMembers(), "expect a cluster of 5")
+
+	var err error
+	var status *Status
+
+	// Test leave handler works correctly
+	status, err = s.testNode.node.adminLeaveHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.Equal(&Status{Status: "ok"}, status)
+	s.Equal(4, s.testNode.node.CountReachableMembers())
+
+	// Test join handler brings it back to 5
+	status, err = s.testNode.node.adminJoinHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.Equal(&Status{Status: "rejoined"}, status)
+	s.Equal(5, s.testNode.node.CountReachableMembers())
+}
+
+// TestRegisterHandlers tests that registerHandler always succeeds.
+func (s *HandlerTestSuite) TestRegisterHandlers() {
+	s.NoError(s.testNode.node.registerHandlers())
+}
+
+func (s *HandlerTestSuite) TestNotImplementedHandler() {
+	_, err := notImplementedHandler(s.ctx, &arg{})
+	s.EqualError(err, "handler not implemented")
+}
+
+// TestErrorHandler tests that the errorHandler logs the correct error message.
+func (s *HandlerTestSuite) TestErrorHandler() {
+	logger := &mocks.Logger{}
+
+	errTest := errors.New("test error")
+
+	logger.On("WithField", "error", errTest).Return(logger)
+	logger.On("Info", []interface{}{"error occurred"})
+
+	originalLogger := s.testNode.node.log
+	s.testNode.node.log = logger
+
+	s.testNode.node.errorHandler(s.ctx, errTest)
+	logger.AssertExpectations(s.T())
+
+	// Restore old logger, as stuff will be logged on teardown
+	s.testNode.node.log = originalLogger
+}
+
+// TestTickHandler tests that the tickHandler is working correctly by stopping
+// all pings and then calling the handler. The handler should trigger a single
+// protocol tick to occur, which means the number of ticks should have
+// incremented by one.
+func (s *HandlerTestSuite) TestTickHandler() {
+	s.testNode.node.Stop()
+	protocolPeriods := s.testNode.node.gossip.protocol.numPeriods
+
+	ping, err := s.testNode.node.tickHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.Equal(protocolPeriods+1, s.testNode.node.gossip.protocol.numPeriods, "protocol period count should have increased by one")
```

Comment:
Can you suggest an alternative method for testing this specific handler that doesn't violate this ideal?

What is the negative impact of violating this in a test?